Plugin: fix Chrome path

AgnesToulet opened the pull request:
The Puppeteer update (https://github.com/grafana/grafana-image-renderer/pull/433) contained an update to the `puppeteer.executablePath` function used in plugin mode to compute Chrome executable path. 

This PR updates the package scripts to ensure Chrome is always installed under `<pluginDir>/chrome` so we can use this path in plugin mode. 

Fixes https://github.com/grafana/grafana-image-renderer/issues/449

Clarity-89 commented on scripts/download_chrome.js:
```diff
@@ -38,4 +42,15 @@ async function download() {
+    // All the chrome files should be in the plugin 'chrome' folder
+    // This is used to set the puppeteer.executablePath
+    const ext = platform === BrowserPlatform.WIN64 ? '.exe' : '';
+    const out = childProcess.execFileSync('find', [`${outputPath}/chrome`, '-type', 'f', '-name', `chrome${ext}`, '-exec', 'dirname', '{}', '\;']);
```
This returns an empty string for me and as a result, the whole build fails. Also I'm not sure I get the purposes of this, before the changes the chrome was being downloaded into `${outputPath}/chrome`, which is the built plugin's folder.

AgnesToulet replied:
The thing is in `app.ts` we need to know where `chrome.exe` is and currently it's in `${outputPath}/chrome/<platform>-<buildVersion>/chrome-<patform>`. The goal of this PR is to get rid of these two intermediate folders that are platform/build-specific so we have a generic path that can be used everywhere. 

Before, we used `executablePath()` to get this information but now that it's not working anymore, I couldn't find another way to do this.

AgnesToulet replied:
The folder structure or the file name may be different on Mac, I'll take a look.

Clarity-89 replied:
Ah ok, I see, makes sense.

Clarity-89 replied:
Actually, I think it's [this one](https://github.com/puppeteer/puppeteer/blob/main/packages/browsers/src/launch.ts#L64).

Clarity-89 commented on scripts/download_chrome.js:
```diff
@@ -38,4 +42,15 @@ async function download() {
+    const chromeBinDir = out.toString().trim()
+    
+    console.log(`Moving ${chromeBinDir} content into ${outputPath}/chrome`);
+    childProcess.execFileSync('mv', [chromeBinDir, `${outputPath}/tmp`]);
```
This command fails when there's no `${outputPath}/tmp` directory.

Clarity-89 commented on scripts/download_chrome.js:
```diff
@@ -25,6 +26,9 @@ const outputPath = path.resolve(process.cwd(), 'dist', process.argv[3] || `plugi
+// Clean up download folder if exists
+childProcess.execFileSync('rm', ['-rf', `${outputPath}/chrome`]);
+
 async function download() {
```
```js
async function download() {
    const buildId = await resolveBuildId(browserVersion, platform, 'latest');
    console.log(`Installing ${browserVersion} into ${outputPath}`);
    const executablePath = computeExecutablePath({buildId, browser: browserVersion, platform, cacheDir: ''});
    const browser = await install({
        cacheDir: outputPath,
        browser: browserVersion,
        platform,
        buildId,
    });

    return {browser, executablePath};
}

download().then(({executablePath}) => {
    console.log(`${browserVersion} downloaded into:`, outputPath);

    // All the chrome files should be in the plugin 'chrome' folder
    // This is used to set the puppeteer.executablePath
    const out = executablePath.substring(0, executablePath.lastIndexOf("/"));
    const chromeBinDir = `${outputPath}/${out}`
    console.log(`Moving ${chromeBinDir} content into ${outputPath}/chrome`);
    childProcess.execFileSync('mv', [chromeBinDir, `${outputPath}/tmp`]);
    childProcess.execFileSync('rm', ['-r', `${outputPath}/chrome`]);
    childProcess.execFileSync('mv', [`${outputPath}/tmp`, `${outputPath}/chrome`]);
});
```

This should work independent of the OS and move the chrome executable directly to the `chrome` folder. The path also needs to be adjusted in `app.ts`. However, this doesn't fix the plugin startup error for me, not sure if it'd be different on Windows.

AgnesToulet replied:
I think this doesn't work on Mac because Chrome executable is not at the root of the Chrome folder as on Linux or Windows. 

I'll look into another solution. I think if we can pass the `buildId` to the package, then we can use the `computeExecutableFunction` to get the whole executable path.

AgnesToulet commented on src/app.ts:
```diff
@@ -78,7 +75,13 @@ async function main() {
+    'error': err.message,
+    'trace': err.stack,
+  }
+  console.error(JSON.stringify(errorLog));
```
This error was actually logged as a "debug" log in plugin mode so that's why it was usually not displayed. This is fixing that and it should help us getting more info from user issues.

AgnesToulet commented:
@Clarity-89 I found another solution, this should be more portable and I find it much cleaner than the other one. Tested on Windows only, I'll do some tests on Linux.

Clarity-89 commented:
> @Clarity-89 I found another solution, this should be more portable and I find it much cleaner than the other one. Tested on Windows only, I'll do some tests on Linux.

Tested on Mac and this works! 🎉  I wonder if this still going to work in docker, since we're still using Chromium in our image.

Clarity-89 commented:
Ok, yeah the Docker image built locally works for me. I'm asking because according to [this issue](https://github.com/grafana/grafana-image-renderer/issues/449) the renderer also fails in Docker on Linux. Is that a separate issue?

Clarity-89 commented on src/app.ts:
```diff
@@ -21,20 +20,18 @@ async function main() {
-        config.rendering.chromeBin = env['PUPPETEER_EXECUTABLE_PATH'];
-      }
+      const execPath = path.dirname(process.execPath);
+      const chromeInfoFile = fs.readFileSync(path.resolve(execPath, 'chrome-info.json'), 'utf8');
+      const chromeInfo = JSON.parse(chromeInfoFile);
```
We'd probably handle the possible errors here and maybe default to executable path from puppeteer (the one in the `.cache` folder)?

Clarity-89 commented on src/app.ts:
```diff
@@ -21,20 +20,18 @@ async function main() {
+      const chromeInfoFile = fs.readFileSync(path.resolve(execPath, 'chrome-info.json'), 'utf8');
+      const chromeInfo = JSON.parse(chromeInfoFile);
+
+      const executablePath = computeExecutablePath({
+        cacheDir: path.dirname(process.execPath),
+        browser: Browser.CHROME,
+        buildId: chromeInfo.buildId,
+      });
+
+      config.rendering.chromeBin = executablePath;
```
```suggestion
     config.rendering.chromeBin = computeExecutablePath({
        cacheDir: path.dirname(process.execPath),
        browser: Browser.CHROME,
        buildId: chromeInfo.buildId,
      });
```

NIT: Don't think an extra variable is necessary here.

AgnesToulet commented:
> I'm asking because according to https://github.com/grafana/grafana-image-renderer/issues/449 the renderer also fails in Docker on Linux. Is that a separate issue?

Yes it's a bit different but also fixed in this PR. In this custom Docker image, the executable path is provided through the `GF_PLUGIN_RENDERING_CHROME_BIN` env variable. But as this variable was used after the executable path computation, it also failed on the call to the `executablePath()` function. 
This is why I moved this section:
```
  if (env['GF_PLUGIN_RENDERING_CHROME_BIN']) {
    config.rendering.chromeBin = env['GF_PLUGIN_RENDERING_CHROME_BIN'];
  }
```
We shouldn't compute the executable path anymore if it's provided.

Clarity-89 reviewed (approved):
Great work on this! 🎉
